### PR TITLE
apiserver: Fix DelegatingAuthenticationOptions temp file leak

### PR DESCRIPTION
**Pull request by sttts**
Follow-up of #42075.

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 1]
```diff
@@ -79,12 +79,14 @@ func (s *RequestHeaderAuthenticationOptions) ToAuthenticationRequestHeaderConfig
 }
 
 type ClientCertAuthenticationOptions struct {
-	// ClientCA is the certificate bundle for all the signers that you'll recognize for incoming client certificates
-	ClientCA string
+	// ClientCAFile is the certificate bundle for all the signers that you'll recognize for incoming client certificates
+	ClientCAFile string
+	// ClientCAData binary equivalent of ClientCAFile. ClientCAFile overrides ClientCAData.
+	ClientCAData []byte
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go]: Shouldn't there be a congruent change for `RequestHeaderAuthenticationOptions`?

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 2]
```diff
@@ -142,12 +144,17 @@ func (s *DelegatingAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 
 }
 
-func (s *DelegatingAuthenticationOptions) ApplyTo(c *server.Config) error {
+func (s *DelegatingAuthenticationOptions) ApplyTo(c *server.Config, certDir string) error {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go]: `certDir` is cruft?

**Comment by deads2k** [comment]
Fix it by taking the proper types down in `DelegatingAuthenticatorConfig`'s transitive structs?

**Comment by deads2k** [comment]
> Fix it by taking the proper types down in DelegatingAuthenticatorConfig's transitive structs?

We might be able to use this one as a starter bug for someone.

**Comment by dchen1107** [comment]
I am moving this to v1.8 since there is no activity for more than a month.

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -367,9 +370,12 @@ func BuildMasterConfig(s *options.ServerRunOptions) (*master.Config, informers.S
 		return nil, nil, err
 	}
 
-	clientCA, err := readCAorNil(s.Authentication.ClientCert.ClientCA)
-	if err != nil {
-		return nil, nil, err
+	clientCA := s.Authentication.ClientCert.ClientCAData
```
Comment (cheftako) [inline cmd/kube-apiserver/app/server.go]: Would seem cleaner to switch readCAorNil to readCAorDefault. Use nil as the default for existing cases and s.Authentication.ClientCert.ClientCAData as the default for this case.

File: pkg/kubeapiserver/options/authentication.go [hunk 4]
```diff
@@ -277,7 +278,16 @@ func (s *BuiltInAuthenticationOptions) ToAuthenticationConfig() authenticator.Au
...
-		ret.ClientCAFile = s.ClientCert.ClientCA
+		if len(s.ClientCert.ClientCAFile) > 0 {
+			var err error
+			ret.ClientCAData, err = ioutil.ReadFile(s.ClientCert.ClientCAFile)
+			if err != nil {
+				return authenticator.AuthenticatorConfig{}, nil
+			}
+
+		} else {
+			ret.ClientCAData = s.ClientCert.ClientCAData
```
Comment (cheftako) [inline pkg/kubeapiserver/options/authentication.go]: Does it make sense if we have both a ca file and passed in ca data to give priority to the file? Better yet can we make sure that both are not set? Why don't we just want them to push the byte array to the FS?

**Pull request closed without merging**